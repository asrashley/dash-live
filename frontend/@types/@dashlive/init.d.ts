import { InitialApiTokens } from "../../src/types/InitialApiTokens";
import { NavBarItem } from "../../src/types/NavBarItem";
import { InitialUserState } from "../../src/types/UserState";

export const initialTokens: InitialApiTokens;

export const navbar: NavBarItem[];

export const user: InitialUserState;