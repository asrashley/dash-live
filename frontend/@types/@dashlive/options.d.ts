import { InputFormGroup } from "../../src/form/types/InputFormGroup";

type OptionMap = {
  [field: string]: boolean | string | number | null;
};

type OptionMapWithChildren = {
  [field: string]: boolean | string | number | null | OptionMap;
};

export const defaultFullOptions: OptionMapWithChildren;

export const defaultShortOptions: OptionMap;

export const defaultCgiOptions: OptionMap;

export const drmSystems: string[];

export const groupNames: string[];

export const fieldGroups: InputFormGroup[];
