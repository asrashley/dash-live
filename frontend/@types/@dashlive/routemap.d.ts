import type { RouteMap } from "../../src/types/RouteMap";

export const routeMap: RouteMap;

export const uiRouteMap: RouteMap;
