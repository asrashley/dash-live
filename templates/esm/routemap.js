export const routeMap = {
{%- for name, route in routes.items() %}
  {{ name }}: {
    title: "{{ route.title }}",
    re: /{{ route.re | safe }}/,
    route: "{{ route.route }}",
    url: {{ route.url | safe }},
  },
{%- endfor %}
};

