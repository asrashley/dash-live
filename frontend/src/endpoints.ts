import { createContext } from 'preact';
import log from 'loglevel';

import { routeMap, uiRouteMap} from '@dashlive/routemap';
import { CsrfTokenStore } from './CsrfTokenStore';
import { CsrfTokenCollection } from './types/CsrfTokenCollection';
import { AllManifests } from './types/AllManifests';
import { AllStreamsJson, AllStreamsResponse } from './types/AllStreams';
import { JWToken } from './types/JWToken';
import { MultiPeriodStream, MultiPeriodStreamJson } from './types/MultiPeriodStream';
import { DecoratedMultiPeriodStream } from "./types/DecoratedMultiPeriodStream";
import { MultiPeriodStreamSummary } from './types/MultiPeriodStreamSummary';
import { ContentRolesMap } from './types/ContentRolesMap';
import { ModifyMultiPeriodStreamJson, ModifyMultiPeriodStreamResponse } from './types/ModifyMultiPeriodStreamResponse';
import { InitialApiTokens } from './types/InitialApiTokens';
import { LoginRequest } from './types/LoginRequest';
import { LoginResponse } from './types/LoginResponse';
import { MultiPeriodStreamValidationRequest, MultiPeriodStreamValidationResponse } from './types/MpsValidation';
import { CgiOptionDescription } from './types/CgiOptionDescription';

type TokenStoreCollection = {
  files: CsrfTokenStore;
  kids: CsrfTokenStore;
  login: CsrfTokenStore;
  streams: CsrfTokenStore;
  upload: CsrfTokenStore;
};

export type ApiRequestOptions = {
  authorization?: string;
  body: RequestInit['body'];
  service: keyof TokenStoreCollection;
  signal?: AbortSignal;
  method: RequestInit['method'];
  rejectOnError?: boolean;
  query?: URLSearchParams;
};

export type GetAllStreamsProps = ApiRequestOptions & {
  withDetails: boolean
};

type RefreshAccessTokenResponse = {
  accessToken?: JWToken;
  csrfTokens: CsrfTokenCollection;
  ok: boolean;
  status: number;
};

export interface ApiRequestsProps extends Readonly<InitialApiTokens> {
  navigate: (url: string) => void;
}

export class ApiRequests {
  private csrfTokens: TokenStoreCollection;
  private accessToken: JWToken | null;
  private refreshToken: JWToken | null;
  private navigate: ApiRequestsProps['navigate'];

  constructor({csrfTokens, accessToken, refreshToken, navigate}: ApiRequestsProps) {
    this.csrfTokens = {
      files: new CsrfTokenStore(csrfTokens?.files),
      kids: new CsrfTokenStore(csrfTokens?.kids),
      login: new CsrfTokenStore(csrfTokens?.login),
      streams: new CsrfTokenStore(csrfTokens?.streams),
      upload: new CsrfTokenStore(csrfTokens?.upload),
    }
    this.accessToken = accessToken;
    this.refreshToken = refreshToken;
    this.navigate = navigate;
  }

  getAllManifests(options: Partial<ApiRequestOptions> = {}): Promise<AllManifests> {
    return this.sendApiRequest<AllManifests>(routeMap.listManifests.url(), options);
  }

  getContentRoles(options: Partial<ApiRequestOptions> = {}): Promise<ContentRolesMap> {
    return this.sendApiRequest(routeMap.contentRoles.url(), options);
  }

  getCgiOptions(options: Partial<ApiRequestOptions> = {}) : Promise<CgiOptionDescription[]> {
    return this.sendApiRequest(routeMap.cgiOptions.url(), options);
  }

  async getUserInfo(signal: AbortSignal): Promise<LoginResponse | Response> {
    if (!this.refreshToken) {
      throw new Error('Cannot get user info without a refresh token');
    }
    const options: Partial<ApiRequestOptions> = {
      authorization: this.refreshToken.jwt,
      method: 'GET',
      rejectOnError: false,
      signal,
    };
    const response = await this.sendApiRequest<LoginResponse | Response>(routeMap.login.url(), options);
    if (response['success']) {
      this.accessToken = (response as LoginResponse).accessToken;
    }
    return response;
  }

  async loginUser(request: LoginRequest, options: Partial<GetAllStreamsProps> = {}): Promise<LoginResponse> {
    const response: LoginResponse = await this.sendApiRequest<LoginResponse>(routeMap.login.url(), {
      body: JSON.stringify(request),
      method: 'POST',
      service: 'login',
      ...options,
    });
    if (response.success) {
      if (response.accessToken) {
        this.accessToken = response.accessToken;
      }
      if (response.refreshToken) {
        this.refreshToken = response.refreshToken;
      }
    }
    return response;
  }

  async logoutUser(options: Partial<GetAllStreamsProps> = {}): Promise<Response> {
    const response: Response = await this.sendApiRequest(routeMap.login.url(), {
      method: 'DELETE',
      service: 'login',
      ...options,
    });
    this.accessToken = null;
    this.refreshToken = null;
    return response;
  }

  async getAllStreams(options: Partial<GetAllStreamsProps> = {}): Promise<AllStreamsResponse> {
    const { streams, keys } = await this.sendApiRequest<AllStreamsJson>(routeMap.listStreams.url(), options);
    return { streams, keys};
  }

  getAllMultiPeriodStreams(options: Partial<ApiRequestOptions> = {}): Promise<MultiPeriodStreamSummary[]> {
    return this.sendApiRequest<MultiPeriodStreamSummary[]>(routeMap.listMps.url(), options);
  }

  async getMultiPeriodStream(mps_name: string, options: Partial<ApiRequestOptions> = {}): Promise<MultiPeriodStream> {
    const { model } = await this.sendApiRequest<MultiPeriodStreamJson>(routeMap.editMps.url({mps_name}), options);
    return model;
  }

  async addMultiPeriodStream(
      data: DecoratedMultiPeriodStream,
      options: Partial<ApiRequestOptions> = {}): Promise<ModifyMultiPeriodStreamResponse> {
    const service = 'streams';
    const csrf_token = await this.csrfTokens[service].getToken(
      options?.signal, this.getCsrfTokens);
    const { errors, success, model } = await this.sendApiRequest<ModifyMultiPeriodStreamJson>(routeMap.addMps.url(), {
      ...options,
      service,
      body: JSON.stringify({
        ...data,
        csrf_token
      }),
      method: 'PUT',
    });
    return { errors, success, model };
  }

  async modifyMultiPeriodStream(mps_name: string, data: DecoratedMultiPeriodStream,
     options: Partial<ApiRequestOptions> = {}): Promise<ModifyMultiPeriodStreamJson> {
    const service = 'streams';
    const csrf_token = await this.csrfTokens[service].getToken(
      options?.signal, this.getCsrfTokens);
    return await this.sendApiRequest<ModifyMultiPeriodStreamJson>(routeMap.editMps.url({mps_name}), {
      ...options,
      service,
      body: JSON.stringify({...data, csrf_token}),
      method: 'POST',
    });
  }

  async deleteMultiPeriodStream(mps_name: string, options: Partial<ApiRequestOptions> = {}): Promise<Response> {
    const service = 'streams';
    const csrf_token = await this.csrfTokens[service].getToken(
      options?.signal, this.getCsrfTokens
    );
    const query = new URLSearchParams({
      csrf_token,
    });
    return this.sendApiRequest<Response>(routeMap.editMps.url({mps_name}), {
      ...options,
      service,
      query,
      method: 'DELETE',
    });
  }

  async validateMultiPeriodStream(data: MultiPeriodStreamValidationRequest,
        options: Partial<ApiRequestOptions> = {}): Promise<MultiPeriodStreamValidationResponse> {
    const service = 'streams';
    const csrf_token = await this.csrfTokens[service].getToken(
      options?.signal, this.getCsrfTokens);
    return this.sendApiRequest<MultiPeriodStreamValidationResponse>(routeMap.validateMps.url(), {
      ...options,
      method: 'POST',
      service: 'streams',
      body: JSON.stringify({...data, csrf_token}),
    });
  }

  private async sendApiRequest<T>(url: string, options: Partial<ApiRequestOptions>): Promise<T> {
    const { authorization, body, service, signal, method='GET',
      rejectOnError = true } = options;
    let { query } = options;
    let usedAccessToken = false;
    const headers: Headers = new Headers({
      "Content-Type": "application/json",
    });
    if (authorization !== undefined) {
      headers.set('Authorization', `Bearer ${authorization}`);
    } else {
      if (!this.accessToken && this.refreshToken) {
        await this.getAccessToken(signal);
      }
      if (this.accessToken) {
        headers.set('Authorization', `Bearer ${this.accessToken.jwt}`);
        usedAccessToken = true;
      } else if (service) {
        const token = await this.csrfTokens[service].getToken(signal);
        if (query === undefined) {
          query = new URLSearchParams({csrf_token: token});
        } else {
          query.set('csrf_token', token);
        }
      }
    }
    if (query) {
      url = `${url}?${query.toString()}`;
    }
    if (signal && signal.aborted) {
      throw signal.reason;
    }
    const cache = 'no-cache';
    const credentials = 'same-origin';
    const mode = 'same-origin';
    url = new URL(url, document.location.origin).href;
    log.trace(`sendApiRequest(${url})`);
    let fetchResult = await fetch(url, {
      cache,
      credentials,
      mode,
      body,
      headers,
      method,
      signal,
    });
    log.trace(`url=${url} status=${fetchResult.status} usedAccessToken=${usedAccessToken}`);
    if (fetchResult.status === 401 && usedAccessToken && this.refreshToken) {
      await this.getAccessToken(signal);
      headers.set('Authorization', `Bearer ${this.accessToken.jwt}`);
      log.trace(`retrying sendApiRequest(${url})`);
      fetchResult = await fetch(url, {
          cache,
          credentials,
          mode,
          body,
          headers,
          method,
          signal,
      });
    }
    if (!fetchResult.ok) {
      if (rejectOnError) {
        throw new Error(`${ url }: ${ fetchResult.status }`);
      } else {
        return fetchResult as T;
      }
    }
    if (fetchResult.status !== 200) {
      return fetchResult as T;
    }
    const data = await fetchResult.json();
    if (signal?.aborted) {
      throw signal.reason;
    }
    if (typeof(data?.csrfTokens) === "object") {
      this.updateCsrfTokens(data.csrfTokens);
    } else if(typeof(data?.csrf_tokens) === "object") {
      this.updateCsrfTokens(data.csrf_tokens);
    } else if (typeof(data?.csrf_token) === "string") {
      this.updateCsrfTokens({[service]: data.csrf_token});
    }

    return data as T;
  }

  private updateCsrfTokens(tokens: Partial<CsrfTokenCollection>) {
    for (const [key, value] of Object.entries(tokens)) {
      if (this.csrfTokens[key] === undefined) {
        this.csrfTokens[key] = new CsrfTokenStore(value);
      } else {
        this.csrfTokens[key].setToken(value);
      }
    }
  }

  private async getAccessToken(signal: AbortSignal): Promise<JWToken> {
    if (!this.refreshToken) {
      throw new Error('Cannot request an access token without a refresh token');
    }
    const options = {
      authorization: this.refreshToken.jwt,
      rejectOnError: false,
      signal,
    };
    const data = await this.sendApiRequest<RefreshAccessTokenResponse>(routeMap.refreshAccessToken.url(), options);
    const { accessToken, csrfTokens, ok, status } = data ?? {};
    if (ok === false && status === 401) {
      this.navigate(uiRouteMap.login.url());
    }
    if (!accessToken) {
      throw new Error('Failed to refresh access token');
    }
    this.accessToken = accessToken;
    if (csrfTokens) {
      this.updateCsrfTokens(csrfTokens);
    }
    return accessToken;
  }

  private getCsrfTokens = async (signal: AbortSignal): Promise<void> => {
    if (!this.accessToken && this.refreshToken) {
      await this.getAccessToken(signal);
    }
    if (!this.accessToken) {
      throw new Error('Cannot request CSRF tokens without an access token');
    }
    const options = {
      authorization: this.accessToken.jwt,
      signal,
    };
    const data = await this.sendApiRequest<RefreshAccessTokenResponse>(routeMap.refreshCsrfTokens.url(), options);
    const { csrfTokens } = data ?? {};
    if (csrfTokens) {
      this.updateCsrfTokens(csrfTokens);
    }
  };
}

export const EndpointContext = createContext<ApiRequests>(null);
