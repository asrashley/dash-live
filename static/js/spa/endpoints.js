import { createContext } from 'preact';
import { routeMap } from '/libs/routemap.js';

class CsrfTokenStore {
  constructor(token) {
    this.token = token;
    this.pending = [];
  }

  setToken(token) {
    if (this.pending.length) {
      const {resolve} = this.pending.shift();
      resolve(token)
    } else {
      this.token = token;
    }
  }

  async getToken(signal) {
    const { token } = this;
    if (token) {
      this.token = null;
      return token;
    }
    const {promise, resolve, reject} = Promise.withResolvers();
    if (signal) {
      signal.addEventListener('abort', () => {
        reject(signal.reason);
      });
    }
    this.pending.push({resolve, reject});
    promise.finally(() => {
      this.pending = this.pending.filter(p => p.resolve !== resolve);
    });
    return await promise;
  }
}

export class ApiRequests {
  constructor({csrfTokens, accessToken, refreshToken}) {
    this.csrfTokens = {
      files: new CsrfTokenStore(csrfTokens?.files),
      kids: new CsrfTokenStore(csrfTokens?.kids),
      streams: new CsrfTokenStore(csrfTokens?.streams),
      upload: new CsrfTokenStore(csrfTokens?.upload),
    }
    this.accessToken = accessToken;
    this.refreshToken = refreshToken;
  }

  getAllStreams({ withDetails = false, ...options} = {}) {
    const query = withDetails ? new URLSearchParams({details:1}) : undefined;
    return this.sendApiRequest(routeMap.listStreams.url(), {
      service: 'streams',
      query,
      ...options,
    });
  }

  getAllMultiPeriodStreams(options) {
    return this.sendApiRequest(routeMap.listMps.url(), {
      service: 'streams',
      ...options,
    });
  }

  getMultiPeriodStream(mps_name, options) {
    return this.sendApiRequest(routeMap.editMps.url({mps_name}), {
      service: 'streams',
      ...options,
    });
  }

  addMultiPeriodStream(data, options) {
    return this.sendApiRequest(routeMap.addMps.url(), {
      ...options,
      service: 'streams',
      body: JSON.stringify(data),
      method: 'PUT',
    });
  }

  modifyMultiPeriodStream(mps_name, data, options) {
    return this.sendApiRequest(routeMap.editMps.url({mps_name}), {
      ...options,
      service: 'streams',
      body: JSON.stringify(data),
      method: 'POST',
    });
  }

  async sendApiRequest(url, options) {
    const { authorization, body, service, signal, method='GET' } = options;
    let { query } = options;
    if (this.csrfTokens[service] === undefined) {
      throw new Error(`Unknown service "${service}"`);
    }
    const headers = {
      "Content-Type": "application/json",
    }
    if (authorization !== undefined) {
      headers.Authorization = `Bearer ${authorization}`;
    } else {
      if (!this.accessToken && this.refreshToken) {
        await this.getAccessToken(signal);
      }
      if (this.accessToken) {
        headers.Authorization = `Bearer ${this.accessToken.jti}`;
      } else {
        const token = await this.csrfTokens[service].getToken();
        if (query === undefined) {
          query = new URLSearchParams({csrf_token: token});
        } else {
          query.set('csrf_token', token);
        }
      }
    }
    if (query) {
      url = `${url}?${query.toString()}`;
    }
    if (signal && signal.aborted) {
      throw signal.reason;
    }
    const fetchResult = await fetch(url, {
      cache: 'no-cache',
      credentials: 'same-origin',
      mode: 'same-origin',
      body,
      headers,
      method,
      signal,
    });
    if (!fetchResult.ok) {
      throw new Error(`Failed to fetch ${ url }: ${ fetchResult.status }`);
    }
    if (signal && signal.aborted) {
      throw signal.reason;
    }
    const data = await fetchResult.json();
    if (typeof(data?.csrfTokens) === "object") {
      this.updateCsrfTokens(data.csrfTokens);
    } else if(typeof(data?.csrf_tokens) === "object") {
      this.updateCsrfTokens(data.csrf_tokens);
    } else if (typeof(data?.csrf_token) === "string") {
      this.updateCsrfTokens({[service]: data.csrf_token});
    }

    return data;
  }

  updateCsrfTokens(tokens) {
    for (const [key, value] of Object.entries(tokens)) {
      if (this.csrfTokens[key] === undefined) {
        this.csrfTokens[key] = new CsrfTokenStore(value);
      } else {
        this.csrfTokens[key].setToken(value);
      }
    }
  }

  async getAccessToken(signal) {
    if (this.refreshToken === undefined) {
      throw new Error('Cannot request an access token without a refresh token');
    }
    const options = {
      authorization: this.refreshToken.jti,
      signal,
    };
    const data = await this.sendApiRequest(routeMap.getAccessToken, {}, options);
    const { accessToken } = data ?? {};
    if (accessToken) {
      this.accessToken = accessToken;
    }
  }
};

export function routeFromUrl(url) {
  for (const [name, route] of Object.entries(routeMap)) {
    if (route.re.test(url)) {
      return {
        name,
        ...route,
      };
    }
  }
  return undefined;
}

export const EndpointContext = createContext(null);
