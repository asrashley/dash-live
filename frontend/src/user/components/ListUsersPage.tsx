import { ReadonlySignal } from "@preact/signals-core";
import { Link } from "wouter-preact";
import { useCallback, useContext } from "preact/hooks";

import { uiRouteMap } from "@dashlive/routemap";
import {
  createSortableTable,
  RenderCellProps,
} from "../../components/SortableTable";
import {
  EditUserState,
  FlattenedUserState,
  useAllUsers,
} from "../../hooks/useAllUsers";
import { BooleanCell } from "../../components/BooleanCell";
import { AppStateContext } from "../../appState";
import { EndpointContext } from "../../endpoints";
import { AddUserDialog } from "./AddUserDialog";

const headings: [keyof FlattenedUserState, string][] = [
  ["pk", "#"],
  ["username", "Username"],
  ["email", "Email"],
  ["lastLogin", "Last Login"],
  ["mustChange", "Must Change"],
  ["adminGroup", "Admin"],
  ["mediaGroup", "Media"],
  ["userGroup", "User"],
];

function renderCell({ field, row }: RenderCellProps<FlattenedUserState>) {
  const { username } = row;
  switch (field) {
    case "username":
      return (
        <Link href={uiRouteMap.editUser.url({ username })}>{username}</Link>
      );
    case "email":
      return (
        <Link href={uiRouteMap.editUser.url({ username })}>{row.email}</Link>
      );
    case "lastLogin":
      return <span>{row.lastLogin ?? "---"}</span>;
    case "mustChange":
    case "adminGroup":
    case "mediaGroup":
    case "userGroup":
      return <BooleanCell value={row[field]} />;
    default:
      return <span>{row[field]}</span>;
  }
}

const UsersTable = createSortableTable<FlattenedUserState>({
  headings,
  primaryKey: "pk",
  initialSortField: "username",
  renderCell,
});

interface ListUsersTableProps {
  users: ReadonlySignal<FlattenedUserState[]>;
}

function ListUsersTable({ users }: ListUsersTableProps) {
  return <UsersTable data={users} caption="Users" />;
}

export default function ListUsersPage() {
  const { dialog } = useContext(AppStateContext);
  const apiRequests = useContext(EndpointContext);
  const { flattenedUsers, validateUser, addUser } = useAllUsers();
  const openAddUserDialog = useCallback(() => {
    dialog.value = {
      backdrop: true,
      addUser: {
        active: true,
      },
    };
  }, [dialog]);
  const closeDialog = useCallback(() => {
    dialog.value = null;
  }, [dialog]);
  const saveNewUser = useCallback(
    async (user: EditUserState) => {
      const errs = validateUser(user);
      if (errs.username) {
        return errs.username;
      }
      if (errs.password) {
        return errs.password;
      }
      try {
        const result = await apiRequests.addUser(user);
        if (result.success) {
          addUser(result.user);
          return "";
        }
        return result.errors.join("\n");
      } catch (err) {
        return `failed to add new user - ${err}`;
      }
    },
    [addUser, apiRequests, validateUser]
  );

  return (
    <div id="user-management" className="container">
      <div className="d-flex flex-row">
        <h1 className="flex-grow-1">User Accounts</h1>
        <button
          onClick={openAddUserDialog}
          className="btn btn-success add-user"
        >
          Add New User
        </button>
      </div>
      <div className="users-list-wrap">
        <ListUsersTable users={flattenedUsers} />
      </div>
      <AddUserDialog
        onClose={closeDialog}
        saveChanges={saveNewUser}
        validateUser={validateUser}
      />
    </div>
  );
}
