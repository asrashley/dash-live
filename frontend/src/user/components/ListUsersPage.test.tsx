import { afterEach, beforeEach, describe, expect, test, vi } from "vitest";
import log from "loglevel";
import fetchMock from "@fetch-mock/vitest";
import userEvent from "@testing-library/user-event";

import { renderWithProviders } from "../../test/renderWithProviders";
import { ApiRequests, EndpointContext } from "../../endpoints";
import ListUsersPage from "./ListUsersPage";

import { routeMap } from "@dashlive/routemap";
import { FakeEndpoint } from "../../test/FakeEndpoint";
import {
  adminUser,
  mediaUser,
  MockDashServer,
  normalUser,
  UserModel,
} from "../../test/MockServer";

describe("ListUsersPage component", () => {
  const needsRefreshToken = vi.fn();
  const hasUserInfo = vi.fn();
  let endpoint: FakeEndpoint;
  let server: MockDashServer;
  let user: UserModel;
  let apiReq: ApiRequests;

  beforeEach(() => {
    log.setLevel('error');
    endpoint = new FakeEndpoint(document.location.origin);
    server = new MockDashServer({
      endpoint,
    });
    user = server.login(adminUser.username, adminUser.password);
    expect(user).not.toBeNull();
    apiReq = new ApiRequests({ needsRefreshToken, hasUserInfo });
    apiReq.setRefreshToken(user.refreshToken);
    apiReq.setAccessToken(user.accessToken);
  });

  afterEach(() => {
    vi.clearAllMocks();
    endpoint.shutdown();
    fetchMock.mockReset();
  });

  test("matches snapshot", async () => {
    const { asFragment, findByText } = renderWithProviders(
      <EndpointContext.Provider value={apiReq}>
        <ListUsersPage />
      </EndpointContext.Provider>
    );
    await findByText(adminUser.email);
    await findByText(mediaUser.email);
    await findByText(normalUser.email);
    expect(asFragment()).toMatchSnapshot();
  });

  test("can add a new user", async () => {
    const user = userEvent.setup();
    const { findByText, findBySelector, getByTestId } = renderWithProviders(
      <EndpointContext.Provider value={apiReq}>
        <ListUsersPage />
      </EndpointContext.Provider>
    );
    const username = "new.username";
    const email = "new.username@test.local";
    const password = "s3cret!";
    const btn = (await findByText("Add New User")) as HTMLButtonElement;
    await user.click(btn);
    await findBySelector(".modal-dialog");
    const userInp = (await findBySelector(
      'input[name="username"]'
    )) as HTMLInputElement;
    await user.clear(userInp);
    await user.type(userInp, username);
    const emailInp = (await findBySelector(
      'input[name="email"]'
    )) as HTMLInputElement;
    await user.clear(emailInp);
    await user.type(emailInp, email);
    const pwdInp = (await findBySelector(
      'input[name="password"]'
    )) as HTMLInputElement;
    await user.clear(pwdInp);
    await user.type(pwdInp, password);
    const confirmInp = (await findBySelector(
      'input[name="confirmPassword"]'
    )) as HTMLInputElement;
    await user.clear(confirmInp);
    await user.type(confirmInp, password);
    const addProm = endpoint.addResponsePromise(
      "put",
      routeMap.listUsers.url()
    );
    await user.click(getByTestId("add-new-user-btn") as HTMLButtonElement);
    const result = await addProm;
    expect(result).toEqual(
      expect.objectContaining({
        status: 200,
        body: expect.any(String),
      })
    );
    const body = JSON.parse(result.body as string);
    const newUser = server.getUser({ username });
    expect(newUser).toBeDefined();
    expect(newUser.password).toEqual(password);
    expect(body).toEqual({
      errors: [],
      success: true,
      user: {
        email,
        username,
        groups: ["USER"],
        lastLogin: null,
        mustChange: true,
        pk: newUser.pk,
      },
    });
  });
});
