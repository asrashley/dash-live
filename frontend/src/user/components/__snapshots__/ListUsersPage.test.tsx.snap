// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ListUsersPage component > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
    id="user-management"
  >
    <div
      class="d-flex flex-row"
    >
      <h1
        class="flex-grow-1"
      >
        User Accounts
      </h1>
      <button
        class="btn btn-success add-user"
      >
        Add New User
      </button>
    </div>
    <div
      class="users-list-wrap"
    >
      <table
        class="table table-striped table-bordered"
        style="width: auto;"
      >
        <caption>
          Users
        </caption>
        <thead>
          <tr
            class="sort_username"
          >
            <th
              class="pk"
            >
              <a
                class=""
                href="#"
              >
                <span
                  class="me-1 text-dark"
                >
                  #
                </span>
              </a>
            </th>
            <th
              class="username"
            >
              <a
                class=" fs-6"
                href="#"
              >
                <span
                  class="me-1 text-dark"
                >
                  Username
                </span>
                <span
                  aria-label="sort-alpha-down icon"
                  class="bi icon bi-sort-alpha-down "
                  role="img"
                />
              </a>
            </th>
            <th
              class="email"
            >
              <a
                class=""
                href="#"
              >
                <span
                  class="me-1 text-dark"
                >
                  Email
                </span>
              </a>
            </th>
            <th
              class="lastLogin"
            >
              <a
                class=""
                href="#"
              >
                <span
                  class="me-1 text-dark"
                >
                  Last Login
                </span>
              </a>
            </th>
            <th
              class="mustChange"
            >
              <a
                class=""
                href="#"
              >
                <span
                  class="me-1 text-dark"
                >
                  Must Change
                </span>
              </a>
            </th>
            <th
              class="adminGroup"
            >
              <a
                class=""
                href="#"
              >
                <span
                  class="me-1 text-dark"
                >
                  Admin
                </span>
              </a>
            </th>
            <th
              class="mediaGroup"
            >
              <a
                class=""
                href="#"
              >
                <span
                  class="me-1 text-dark"
                >
                  Media
                </span>
              </a>
            </th>
            <th
              class="userGroup"
            >
              <a
                class=""
                href="#"
              >
                <span
                  class="me-1 text-dark"
                >
                  User
                </span>
              </a>
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            id="opt_1"
          >
            <td
              class="pk"
            >
              <span>
                1
              </span>
            </td>
            <td
              class="username"
            >
              <a
                href="/users/admin"
              >
                admin
              </a>
            </td>
            <td
              class="email"
            >
              <a
                href="/users/admin"
              >
                admin@example.local
              </a>
            </td>
            <td
              class="lastLogin"
            >
              <span>
                ---
              </span>
            </td>
            <td
              class="mustChange"
            >
              <span
                class="bool-no"
              >
                ✗
              </span>
            </td>
            <td
              class="adminGroup"
            >
              <span
                class="bool-yes"
              >
                ✓
              </span>
            </td>
            <td
              class="mediaGroup"
            >
              <span
                class="bool-yes"
              >
                ✓
              </span>
            </td>
            <td
              class="userGroup"
            >
              <span
                class="bool-yes"
              >
                ✓
              </span>
            </td>
          </tr>
          <tr
            id="opt_101"
          >
            <td
              class="pk"
            >
              <span>
                101
              </span>
            </td>
            <td
              class="username"
            >
              <a
                href="/users/mediamgr"
              >
                mediamgr
              </a>
            </td>
            <td
              class="email"
            >
              <a
                href="/users/mediamgr"
              >
                media.manager@example.local
              </a>
            </td>
            <td
              class="lastLogin"
            >
              <span>
                ---
              </span>
            </td>
            <td
              class="mustChange"
            >
              <span
                class="bool-no"
              >
                ✗
              </span>
            </td>
            <td
              class="adminGroup"
            >
              <span
                class="bool-no"
              >
                ✗
              </span>
            </td>
            <td
              class="mediaGroup"
            >
              <span
                class="bool-yes"
              >
                ✓
              </span>
            </td>
            <td
              class="userGroup"
            >
              <span
                class="bool-yes"
              >
                ✓
              </span>
            </td>
          </tr>
          <tr
            id="opt_100"
          >
            <td
              class="pk"
            >
              <span>
                100
              </span>
            </td>
            <td
              class="username"
            >
              <a
                href="/users/user"
              >
                user
              </a>
            </td>
            <td
              class="email"
            >
              <a
                href="/users/user"
              >
                a.user@example.local
              </a>
            </td>
            <td
              class="lastLogin"
            >
              <span>
                ---
              </span>
            </td>
            <td
              class="mustChange"
            >
              <span
                class="bool-no"
              >
                ✗
              </span>
            </td>
            <td
              class="adminGroup"
            >
              <span
                class="bool-no"
              >
                ✗
              </span>
            </td>
            <td
              class="mediaGroup"
            >
              <span
                class="bool-no"
              >
                ✗
              </span>
            </td>
            <td
              class="userGroup"
            >
              <span
                class="bool-yes"
              >
                ✓
              </span>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</DocumentFragment>
`;
